tests/MongoDB.Driver.Tests/Specifications/Runner/MongoClientJsonDrivenTestRunnerBase.cs

```diff
@@ -0,0 +1,392 @@
+            }
+        }
+
+        protected virtual void ConfigureClientSettings(MongoClientSettings settings, BsonDocument test)
+        {
+            if (test.Contains("clientOptions"))
+            {
+                foreach (var option in test["clientOptions"].AsBsonDocument)
+                {
+                    switch (option.Name)
+                    {
+                        case "readConcernLevel":
+                            var level = (ReadConcernLevel)Enum.Parse(typeof(ReadConcernLevel), option.Value.AsString, ignoreCase: true);
+                            settings.ReadConcern = new ReadConcern(level);
+                            break;
+
+                        case "readPreference":
+                            settings.ReadPreference = ReadPreferenceFromBsonValue(option.Value);
+                            break;
+
+                        case "retryWrites":
+                            settings.RetryWrites = option.Value.ToBoolean();
+                            break;
+
+                        case "w":
+                            if (option.Value.IsString)
+                            {
+                                settings.WriteConcern = new WriteConcern(option.Value.AsString);
+                            }
+                            else
+                            {
+                                settings.WriteConcern = new WriteConcern(option.Value.ToInt32());
+                            }
+                            break;
+
+                        default:
+                            throw new FormatException($"Unexpected client option: \"{option.Name}\".");
+                    }
+                }
+            }
+        }
+
+        protected virtual void CustomDataValidation(BsonDocument shared, BsonDocument test)
+        {
+            // do nothing by default.
+        }
+
+        protected virtual void ExecuteOperations(IMongoClient client, Dictionary<string, object> objectMap, BsonDocument test)
+        {
+            _objectMap = objectMap;
+
+            var factory = new JsonDrivenTestFactory(client, DatabaseName, CollectionName, objectMap);
+
+            foreach (var operation in test[OperationsKey].AsBsonArray.Cast<BsonDocument>())
+            {
+                var receiver = operation["object"].AsString;
+                var name = operation["name"].AsString;
+                var jsonDrivenTest = factory.CreateTest(receiver, name);
+
+                jsonDrivenTest.Arrange(operation);
+                if (test["async"].AsBoolean)
+                {
+                    jsonDrivenTest.ActAsync(CancellationToken.None).GetAwaiter().GetResult();
+                }
+                else
+                {
+                    jsonDrivenTest.Act(CancellationToken.None);
+                }
+                jsonDrivenTest.Assert();
+            }
+        }
+
+        protected abstract void TestBody(BsonDocument shared, BsonDocument test, EventCapturer eventCapturer);
```

Comment:
Let's name this method:

    RunTest

which will be unique after we rename the other RunTest methods (see other comments).

Reply:
Done